README.MD
RealtimeMonitor v1.1.1
======================

See LICENSE for this software's licensing terms.  At first glance, it resembles the BSD-3 license, but it contains a significant difference which prohibits certain use.  Read it carefully.  Licenses for third-party components used by RealtimeMonitor are in THIRD_PARTY_LICENSES.MD.

This is a pure JavaScript application (no jQuery, etc.) that is used for monitoring numeric data returned by a server.  The data could be from anything at all - machinery, atmospheric sensors... any type of numeric data that you can get into a server is fair game.

Supported methods of data retrieval are Web sockets and AJAX GET/POST polling.  The server response must be JSON.

Knowing what the server response looks like, you create a corresponding configuration object to set up the fields you want to display.  Optional high and low thresholds can be set on each field.  Each threshold has a warning level and a danger level.  Visual feedback is given if the data falls outside of the established boundaries.

The fields are laid out vertically in a panel, and you can create as many panels as you wish.  Current, highest and lowest values are tracked, and realtime graphs display the last 35 data points for each field.

Compatibility is limited to browsers which support [const](https://caniuse.com/#search=const), [let](https://caniuse.com/#search=let) and [mutation observers](https://caniuse.com/#search=mutationobserver), which means anything loosely modern, including the aging IE 11.


## Using

Using RealtimeMonitor is pretty simple - there are three steps:

1.  Instantiate RealtimeMonitor
2.  Create a panel via the `newPanel` function
3.  Append the returned DOM object to something

```javascript
var rtm = new RealtimeMonitor();
var panel = rtm.newPanel( options );
element.appendChild( panel );
```

See the `demo` directory for in-depth code examples of various panel configurations.

### Options

The configuration object passed into `newPanel` has the following properties:

| Property  | Data Type | Description   | Required ? | Default Value |
|-----------|-----------|---------------|------------|---------------|
| `title`   | string    | Text displayed in the panel's title bar | yes | none
|[`url`](#url-property)| object  | Specifies how data is retrieved.  Click the property name to be taken to the definition of this nested object. | yes | none
|`autoConnect`| boolean | If `true`, the panel will connect as soon as it's added to the page | no | true
|`startMinimized`|boolean | If `true`, the panel initializes collapsed down to its title bar | no | false
|`notifications`| boolean | If `true`, and if the browser supports it, native system notifications will display when low or high thresholds reach the warning or danger level | no | true
|[`fields`](#fields-property)|object array| Specifies field configuration.  Each element of the array contains the configuration for a single field.  Click the property name to be taken to the definition of this nested object.| yes | none


#### `url` property
| Property   | Data Type | Description   | Required ? | Default Value |
|------------|-----------|---------------|------------|---------------|
| `address`  | string    | The URL used to update the panel.  Supported protocols are http, https, ws and wss. | yes | none
| `method`   | string    | The HTTP method.  Valid values are **GET** and **POST**.  This property is not used with Web sockets. | yes, if using HTTP(S) | none
| `postData` | object    | Only define this property if `method` is **POST**.  An object consisting of key/value pairs defining fields to be sent in the POST. | yes, if POST | none
|`pollInterval`| int     | The GET/POST polling interval, in seconds.  The lowest possible value is 3; if the value is less than 3, it is ignored and 3 is used instead.  (Not used with Web sockets - Web sockets have updates pushed to them whenever the server decides to).  | no | 3
|`wsGreeting`| object    | For use with Web sockets only.  If for whatever reason you need to send data as soon as the connection is established, provide it here.  The object will be transmitted as JSON, and will be transmitted only once. | no | none
|`wsCloseCodes`| int[ ]  | For use with Web sockets only.  Specifies [CloseEvent.code values](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Properties) that, if encountered, will result in the panel attempting to automatically reconnect.  1000 (normal closure) is not allowed; using it will cause initialization to abort.  You'll need a carefully-coded back end to be sure of the reason you're reconnecting. | no | none

[Back to Options](#options)


#### `fields` property
| Property  | Data Type | Description   | Required ? | Default Value |
|-----------|-----------|---------------|------------|---------------|
| `prop`    | string    | The property name in the JSON response | yes | none
| `label`   | string    | Text displayed in front of the value | yes | none
| `suffix`  | string    | Text displayed after the value to indicate unit of measurement, etc. | no | none
| [`lowThresholds`](#lowthresholdshighthresholds-property) | object| An optional object specifying numeric warning and danger levels - used to drive visual feedback.  Click the property name to be taken to the definition of this nested object. | no | none
| [`highThresholds`](#lowthresholdshighthresholds-property)|object | Same idea as `lowThresholds` | no | none
| `showLowest`| boolean | Specifies whether to display the lowest recorded value for `prop`.  When enabled, the value will appear as a separate field immediately beneath the field being tracked.  When disabled, the value will still be viewable as a tooltip on the field name. | no | true
|`showHighest`| boolean | Same idea as `showLowest` | no | true

[Back to Options](#options)


#### `lowThresholds/highThresholds` property
| Property  | Data Type | Description   | Required ? | Default Value |
|-----------|-----------|---------------|------------|--------------|
| `warn`    | number    | The warning threshold | Only if `danger` is not defined | none
| `danger`  | number    | The danger threshold  | Only if `warn` is not defined | none

[Back to Fields](#fields-property)

[Back to Options](#options)


## Public Methods

| Method                 | Return Type  | Description                        |
|------------------------|--------------|------------------------------------|
| `getAvailableThemes()` | string[ ]    | Returns all available themes.  The array contains the `title` attribute of all alternate stylesheets. |
| `getTheme()`           | string       | Returns the name of the current theme.  This is the `title` attribute of the currently active alternate stylesheet.  Returns null if no theme is in use. |
| `loadTheme(string)`    | void         | Changes the current theme - in other words, sets an alternate stylesheet to active based on a `title` attribute matching the supplied argument. |


## Events

| Event           | Description                                              |
|-----------------|----------------------------------------------------------|
| RTMThemeChange* | Fired when the theme is changed.  The theme name is attached to the event in `event.detail.themeName`. |

\* For IE 11 compatibility, RealtimeMonitor contains a polyfill for CustomEvent.  If your practice is to keep polyfills separate in your projects, you'll want to remove it from your copy of RealtimeMonitor.


## Running the Demo/Development Environment

RealtimeMonitor doesn't have an offline simulator.  To see it process data, you'll need a Web application that returns JSON in response to requests.

In the `demo` directory you'll find a barebones Node.js server which does this.  [Install Node.js](https://nodejs.org) and OpenSSL if you don't have them, then:

For Windows users, run `startserver.bat`.  The batch script takes care of dependencies, such installing required NPM modules and creating a certificate.

If not using `startserver.bat`, change into the **base** directory, then install the server dependencies:

```shell
$ cd js-RealtimeMonitor
$ npm install ws
```

Then, without changing the working directory:

```shell
$ node demo/server.js
```

By design, the server only honors requests from localhost.

Once the server is running, open `demo/demo.html` straight from your filesystem, and it will automatically redirect to the secure localhost URL.


## Color Coding

Assuming you've declared thresholds for a field, each status (normal, warning and danger) has a different color associated with it.  Color coding is used in three ways:

1. The status of the latest reading is represented as a colored square to the left of the field.  The status is determined by comparing the value to the high and low thresholds.

2. The highest & lowest (the historical) readings are displayed as colored text.  These values are also compared against the high and low thresholds.  Unlike the real-time reading, these fields only change color if the values deteriorate to the point that a worse threshold is breached.  In other words, once it reaches 'warning', it can't get back to 'normal', and once it reaches 'danger', it remains there forever.

3.  The color of the titlebar indicates the worst status among the latest readings of all fields.


## Graphs

At the bottom of each panel is an area that contains a line graph for the currently chosen field.  Cycle through the graphs by clicking a field label.

RealtimeMonitor uses [Chart.js](http://chartjs.org) for graphing.  Compatibility has only been verified with Chart.js v2.7.2, the latest available at the time this software was written.  Using a different version could potentially cause a theme's graph colors to malfunction - you'll have to test that yourself.

The graph configurations have been tuned for readability and performance:

1.  Graphs display a maximum of 35 points at a time, after which data points get shifted off the left edge.
2.  All Chart.js animations are disabled.  In Firefox on a Core i7 with one panel containing 4 fields, CPU use often reached 15% with animations enabled (!).  Granted, Firefox isn't the most performant browser out there, but this benchmark wasn't acceptable.  With animations disabled, as well as a few other tweaks, the same test scenario resulted in CPU use of 0-1%.

I contemplated making the Chart.js settings configurable so that these choices are not forced on everyone, but ultimately I decided against it.  I don't recommend changing these settings.


## Notifications

The application can generate system notifications when a warning or danger threshold is reached.  This feature is enabled/disabled through configuration.  The goal behind adding notification support is that you can keep your browser out of view and still know when something requires your attention.

Notification technology is currently evolving.  As of early 2018, it's still somewhat of a mess; there are deprecated APIs and browsers differ greatly in supported capabilities.  IE has no support; Firefox and Chrome support them, but each browser has its own set of annoyances - more on that in a moment.

Each panel displays a notification when one of its warning or danger thresholds is reached.  Because it's important to not miss one of these events, the code is written so that the notifications are persistent:  they won't close on their own - the user must close them.  A panel cannot have more than one open notification at a time, so unless you have a dozen panels and conditions go south in all of them, you won't get buried in notifications.

There *is* a possible annoyance factor, though.  Every time a panel refreshes, it evaluates whether any of its readings crosses into the warning/danger zone, and if so, it displays a notification.  That means that if the user closes a notification, it can pop right back up the next time the panel refreshes.  Since this could be rather infuriating, the application provides a menu option within each panel to mute/unmute notifications from that panel.  As was mentioned at the beginning of this section, notifications can also be defaulted to off via configuration.

Now, onto the browser-specific issues.

### Firefox
By default, Firefox ignores the configuration option that makes notifications persistent.  See [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1346375).  To enable persistent notifications, *dom.webnotifications.requireinteraction.enabled* must be set to `true` in about:config.  So, unless you/your users are going to be staring at panels continuously, or unless you can get them to fiddle with browser configuration, Firefox is probably a poor choice.

### Chrome
Chrome only displays notifications if the page is loaded over HTTPS - or so I've read.  If that's true, there appears to be an exemption for servers running off of localhost, because I ran it on an insecure local server just fine.


## Themes

The application includes several alternate color schemes, or themes.  Themes are implemented by alternate stylesheets which override portions of the base stylesheet.  Some browsers, such as Firefox and IE, place alternate stylesheets in a menu that the user can then select from to switch to a different theme.  Chrome does not have built-in support for this functionality.  I've read that there are third-party Chrome addons that add this capability, but I haven't looked into it any further than that.

Chromium (and by extension, all Chromium-based browsers) has a bug where alternate stylesheets aren't correctly initialized upon page load.  Chromium initializes the 'disabled' property of every alternate stylesheet to `false`, which is wrong.  The workaround for this needs to be made at the page level, not within RealtimeMonitor.  If you plan on using themes, you need to forcibly set this property to `true` on all alternate stylesheets before instantiating RealtimeMonitor.  See the `fixBrokenBrowsers` function in `demo.js` for code which does this.


## Third-party Attribution

1. Graphs are implemented with [Chart.js](http://www.chartjs.org)

2. The [warning](https://pixabay.com/en/sign-triangle-caution-yellow-36071) and [danger](https://pixabay.com/en/sign-triangle-attention-warning-36070) notification icons were provided by Pixabay user [Clker-Free-Vector-Images](https://pixabay.com/en/users/Clker-Free-Vector-Images-3736).


## Donations

https://paypal.me/KurtisLoVerde/10

Thank you for your support!


CHANGELOG.md
# Release 1.1.1 (May 13, 2018)

* Updated the demo to handle a Chromium bug that affects the proper operation of themes.  See README.MD and demo.js for more information.


# Release 1.1 (May 9, 2018)

This release improves theme support.

* Added public method `getAvailableThemes` for getting available themes
* Added public method `loadTheme` for changing the current theme
* An `RTMThemeChange` event now fires when RealtimeMonitor detects a theme change
* Updated the demo to incorporate the new features of this release


# Release 1.0 (April 26, 2018)

* First release

demo/demo.js
/*
 * RealtimeMonitor
 * https://www.github.com/kloverde/js-RealtimeMonitor
 *
 * Copyright (c) 2018, Kurtis LoVerde
 * All rights reserved.
 *
 * Donations:  https://paypal.me/KurtisLoVerde/10
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     1. This software may not be used by any for-profit entity, whether a
 *        business, person, or other, or for any for-profit purpose.
 *     2. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *     3. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *     4. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

"use strict";

document.addEventListener( "DOMContentLoaded", function() {
   fixBrokenBrowsers();  // You should read the documentation in this function.

   const rtm = new RealtimeMonitor();

   const panel1 = rtm.newPanel( {
      title  : "GET Polling",
      url    : { address : "https://localhost:8081/status/site1",
                 method  : "GET",
                 pollInterval : 4 },
      notifications : false,
      fields : [
                  { prop : "load", label : "Load", suffix : "%", highThresholds : { warn : 70, danger : 90 } },
                  { prop : "rpm", label : "RPM",  lowThresholds : { warn : 500, danger : 300 }, highThresholds : { warn : 2000, danger : 2500 } },
                  { prop : "ambientTemp", label : "Ambient Temp", suffix : "°F", showLowest : false, showHighest : false },
                  { prop : "internalTemp", label : "Internal Temp", suffix : "°F", highThresholds : { warn : 200, danger : 250 } }
               ]
   } );

   const panel2 = rtm.newPanel( {
      title  : "POST Polling",
      url    : { address  : "https://localhost:8081/status",
                 method   : "POST",
                 postData : { site : "site2", otherStuff : "more_stuff" } },
      autoConnect    : true,
      startMinimized : false,
      notifications  : false,
      fields : [
                  { prop : "load", label : "Load", suffix : "%", highThresholds : { warn : 70, danger : 90 }, showLowest : true, showHighest : true },
                  { prop : "rpm", label : "RPM",  lowThresholds : { warn : 500, danger : 300 }, highThresholds : { warn : 2000, danger : 2500 }, showLowest : true, showHighest : true },
                  { prop : "ambientTemp", label : "Ambient Temp", suffix : "°F", showLowest : false, showHighest : false },
                  { prop : "internalTemp", label : "Internal Temp", suffix : "°F", highThresholds : { warn : 200, danger : 250 }, showHighest : true }
               ]
   } );

   const panel3 = rtm.newPanel( {
      title  : "Web Socket",
      url    : { address      : "wss://localhost:8081/status/site1",
                 wsGreeting   : { stuff : "hey_stuff" },
                 wsCloseCodes : [ 1006, 1013 ] },
      autoConnect    : true,
      startMinimized : false,
      notifications  : false,
      fields : [
                  { prop : "load", label : "Load", suffix : "%", highThresholds : { warn : 70, danger : 90 }, showLowest : true, showHighest : true },
                  { prop : "rpm", label : "RPM",  lowThresholds : { warn : 500, danger : 300 }, highThresholds : { warn : 2000, danger : 2500 }, showLowest : true, showHighest : true },
                  { prop : "ambientTemp", label : "Ambient Temp", suffix : "°F", showLowest : false, showHighest : false },
                  { prop : "internalTemp", label : "Internal Temp", suffix : "°F", highThresholds : { warn : 200, danger : 250 }, showHighest : true }
               ]
   } );

   // You'll have to deactivate the visibilityHidden class to observe the XSS3 test, as suffixes are hidden until the 'Connect' menu option is clicked

   const panel4 = rtm.newPanel( {
      title  : "Test XSS Attacks <img src=\"asdf\" onerror=\"alert('xss1')\" />",
      url    : { address  : "https://localhost:8081/status/site1",
                 method   : "GET" },
      autoConnect    : false,
      startMinimized : false,
      notifications  : false,
      fields : [
                  { prop : "xss2", label : "XSS 2 <img src=\"asdf\" onerror=\"alert('xss2')\" />", suffix : "suffix", showHighest : false },
                  { prop : "xss3", label : "XSS 3", suffix : "suffix <img src=\"asdf\" onerror=\"alert('xss3')\" />", showHighest : false },
                  { prop : "jsonXss", label : "JSON XSS Test (click 'Connect')", showHighest : false }
               ]
   } );

   document.addEventListener( "RTMThemeChange", function(event) {
      saveThemeToCookie( event.detail.themeName );
   } );

   loadThemeFromCookie();

   document.body.appendChild( panel1 );
   document.body.appendChild( panel2 );
   document.body.appendChild( panel3 );
   document.body.appendChild( panel4 );

   function saveThemeToCookie( themeName ) {
      const expires = new Date();
		expires.setTime( expires.getTime() + (365 * 24 * 60 * 60 * 1000) );

      // Could use rtm.getTheme() to get the current theme, but that's not as efficient when
      // you've already received the theme name from a RTMThemeChange CustomEvent.
      document.cookie = "rtmTheme=" + encodeURIComponent(themeName) + "; expires=" + expires.toGMTString();
   }

   function loadThemeFromCookie() {
      const theme = decodeURIComponent( document.cookie ).substr( 9 );

      if( ("" + theme).length > 0 ) {
         rtm.loadTheme( theme );
      }
   }

   function fixBrokenBrowsers() {
      // Forcibly set all alternate stylesheets to disabled.  You MUST do this if you're using RealtimeMonitor's themes in your project.
      // There's a bug in Chromium where all alternate stylesheets are initialized with their 'disabled' property set to false, even
      // though they all start out disabled.  This is incorrect behavior, and it interferes with how RealtimeMonitor detects the active
      // theme.  This affects all Chromium browsers, so Chrome, Brave, etc. will not function correctly unless you do this.

      const sheets = document.querySelectorAll( "link[rel='alternate stylesheet']" );

      for( let i = 0; i < sheets.length; i++ ) {
         sheets[i].disabled = true;
      }
   }
} );


front_ends/plain_javascript/js/realtime_monitor.js
/*
 * RealtimeMonitor
 * https://www.github.com/kloverde/js-RealtimeMonitor
 *
 * Copyright (c) 2018, Kurtis LoVerde
 * All rights reserved.
 *
 * Donations:  https://paypal.me/KurtisLoVerde/5
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     1. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *     2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *     3. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

"use strict";

// The UI is created dynamically from a supplied array of objects.  Each array element specifies the configuration
// of an individual panel; you need to define at least one, and there is no upper limit.  The object has the
// following members:
//
// title :  Text displayed at the top of the panel
// url   :  The URL used to update the panel
// fields:  An array of objects specifying field configuration:
//             prop       : The property name present in the JSON response - also used in the HTML ID
//             label      : Text displayed in front of the value
//             suffix     : Text displayed after the value.  Optional.
//             thresholds : An optional object specifying warning and danger levels - used to drive visual feedback.
//                          Supports only numeric values, so if your data is text you'll have to map it to a number.
//                            warn   : The warning threshold
//                            danger : The danger threshold
//             showMax    : A boolean which specifies whether to display the largest recorded value for 'prop'.
//                          Assumes numeric values, so if your data is text you'll have to map it to a number.
//                          When set to true, the max value will appear as a separate field immediately beneath the
//                          field being tracked.

function RealtimeMonitor() {
   var PREFIX_MAX = "max",
       PREFIX_PANEL = "panel";

   var thresholds = [];
   var panelData = [];

   this.buildUI = function( appCfg ) {
      for( var i = 0; i < appCfg.length; i++ ) {
         var panelCfg = appCfg[i];

         var panelContainer = document.createElement( "div" );
         var title = document.createElement( "div" );
         var btnContainer = document.createElement( "div" );
         var btn = document.createElement( "button" );
         var btnValue = document.createTextNode( "Connect" );
         var graphContainer = document.createElement( "div" );
         var graphs = [];

         panelContainer.id = PREFIX_PANEL + i;
         panelContainer.className = "monitoringPanel";

         title.id = "title" + i;
         title.className = "title";
         title.appendChild( document.createTextNode(panelCfg.title) );

         panelContainer.appendChild( title );

         thresholds[i] = [];
         panelData[PREFIX_PANEL + i] = panelData[PREFIX_PANEL + i] || [];

         for( var j = 0; j < panelCfg.fields.length; j++ ) {
            var fieldCfg = panelCfg.fields[j];

            if( fieldCfg.thresholds ) {
               thresholds[i][fieldCfg.prop] = fieldCfg.thresholds;  // We need to refer to this part of the configuration after the UI is built, so save it
            }

            panelContainer.appendChild( newField(false, fieldCfg.prop, i, Boolean(fieldCfg.thresholds), fieldCfg.label, fieldCfg.suffix) );

            if( fieldCfg.showMax ) {
               panelContainer.appendChild( newField(true, fieldCfg.prop, i, Boolean(fieldCfg.thresholds), fieldCfg.label, fieldCfg.suffix) );
            }

            panelContainer.appendChild( newFieldSeparator() );
            panelData[PREFIX_PANEL + i][PREFIX_MAX + fieldCfg.prop] = 0;

            graphs.push( newGraph(fieldCfg.prop, i) );
         }

         graphContainer.className = "graphContainer";
         panelContainer.appendChild( graphContainer );

         btnContainer.className = "btnContainer";

         btn.id = "btn" + i;
         btn.addEventListener( "click", function(event) { btnClick( this ) } );
         btn.appendChild( btnValue );
         btnContainer.appendChild( btn );

         panelContainer.appendChild( btnContainer );

         for( var j = 0; j < graphs.length; j++ ) {
            graphContainer.appendChild( graphs[j] );
         }

         document.body.appendChild( panelContainer );

         function newField( isMax, propName, panelNum, hasThreshold, labelText, suffix ) {
            var fieldContainer = document.createElement( "div" );
            var status = document.createElement( "div" );
            var label = document.createElement( "label" );
            var val = document.createElement( "span" );

            fieldContainer.className = "fieldContainer";

            if( isMax ) {
               propName = PREFIX_MAX + propName;
               labelText = "Max " + labelText;
            } else {
               fieldContainer.className += " hasGraph";

               label.addEventListener( "click", function(event) {
                  showGraph( "graph" + this.getAttribute("for") );
               } );
            }

            status.id = "status" + propName + panelNum;  // Display color coding when there's an ID, otherwise keep the element for the sake of consistent indentation
            status.className = "status"
            fieldContainer.appendChild( status );

            val.id = propName + panelNum;

            label.setAttribute( "for", val.id );
            label.appendChild( document.createTextNode(labelText) );
            fieldContainer.appendChild( label );

            fieldContainer.appendChild( val );

            if( suffix ) {
               fieldContainer.appendChild( newSuffix(suffix, propName, panelNum) );
            }

            return fieldContainer;
         }

         function newSuffix( suffix, prop, panelNum ) {
            var elem = document.createElement( "span" );
            elem.id = "suffix" + prop + panelNum;
            elem.className = "hidden";
            elem.appendChild( document.createTextNode(suffix) );
            return elem;
         }

         function newFieldSeparator() {
            var separator = document.createElement( "div" );
            separator.className = "fieldSeparator";
            return separator;
         }

         function newGraph( propName, panelNum ) {
            var graph = document.createElement( "div" );

            graph.id = "graph" + propName + panelNum;
            graph.appendChild( document.createTextNode(graph.id) );

            return graph;
         }
      }
   };

   function showGraph( id ) {
      var graph = document.getElementById( id );
      var graphContainer = graph.parentNode;

      for( var i = 0; i < graphContainer.childNodes.length; i++ ) {
         graphContainer.childNodes[i].style.visibility = "hidden";
      }

      graph.style.visibility = "visible";
   }

   function btnClick( btn ) {
      var panelNum = btn.id.replace( "btn", "" );
      var connected = btn.innerHTML.indexOf( "Disconnect" ) !== -1;

      if( connected ) {
         disconnect( panelNum );
         btn.innerHTML = "Connect";
      } else {
         connect( panelNum );
         btn.innerHTML = "Disconnect";
      }
   }

   var simulator = null;

   function connect( panelNum ) {
      simulator = window.setInterval( function() {
         var jsonResponse = JSON.stringify( {
            load : random(50, 100),
            //load : "<img src=\"asdf\" onerror=\"console.log('json xss')\" />",
            rpm  : random(1500, 2700),
            ambientTemp  : random(70, 75),
            internalTemp : random(175, 260),
            rhinocerous  : 45  // unrecognized properties do not cause errors
         } );

         updateStats( panelNum, jsonResponse );
         updateUI( panelNum );
      }, 1000 );

      function random( from, to ) {
         return Math.floor( Math.random() * (to - from + 1) ) + from;
      }
   }

   function disconnect( panelNum ) {
      window.clearInterval( simulator );
   }

   function updateStats( panelNum, jsonResponse ) {
      var panel = panelData[ PREFIX_PANEL + panelNum ];
      var stats = JSON.parse( jsonResponse );

      for( var prop in stats ) {
         var maxProp = PREFIX_MAX + prop;

         // Kind of a kludge.  To save on memory, the configuration passed to buildUI isn't saved, so we
         // use the presence of a "max" property to see if the property from the server is recognized.

         if( panel[maxProp] !== undefined ) {
            panel[prop] = stats[prop];
            panel[maxProp] = panel[maxProp] >= panel[prop] ? panel[maxProp] : panel[prop];
         }
      }
   }

   function updateUI( panelNum ) {
      var panelThresholds = thresholds[panelNum];
      var data = panelData[ PREFIX_PANEL + panelNum ];

      for( var prop in data ) {
         var thresholdProp = prop.replace( new RegExp("^" + PREFIX_MAX), "" );
         var value = data[prop];
         var field = document.getElementById( prop + panelNum );

         if( field != null ) {
            var className = "normal";
            var fieldThresholds = panelThresholds[thresholdProp];
            var status = document.getElementById( "status" + prop + panelNum );

            field.innerHTML = "";
            field.appendChild( document.createTextNode(value) );

            if( fieldThresholds ) {
               if( fieldThresholds.danger && value >= fieldThresholds.danger ) {
                  className = "danger";
               } else if( fieldThresholds.warn && value >= fieldThresholds.warn ) {
                  className = "warn";
               }
            }

            status.className = "status " + (fieldThresholds ? className : "none");

            var suffix = document.getElementById( "suffix" + prop + panelNum );

            if( suffix ) {
               suffix.className = "";  // make visible
            }
         }
      }
   }
}
